pkg/monitor/monitor/monitor_manager.go

```diff
@@ -427,3 +422,253 @@ func (m *MonitorManager) patchTidbClusterStatus(tcRef *v1alpha1.TidbClusterRef,
 	return err
 }
+
+func (m *MonitorManager) smoothMigrationToStatefulSet(monitor *v1alpha1.TidbMonitor) (bool, error) {
+	// determine whether there is an old deployment
+	oldDeploymentName := GetMonitorObjectName(monitor)
+	oldDeployment, err := m.deps.DeploymentLister.Deployments(monitor.Namespace).Get(oldDeploymentName)
+	if err == nil {
+		klog.Infof("The old deployment exists, start smooth migration for tm [%s/%s]", monitor.Namespace, monitor.Name)
+		// if deployment exist, delete it and wait next reconcile.
+		err = m.deps.TypedControl.Delete(monitor, &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      oldDeployment.Name,
+				Namespace: oldDeployment.Namespace,
+			},
+		})
+		if err != nil {
+			klog.Errorf("Smooth migration for tm[%s/%s], fail to delete the old deployment, err: %v", monitor.Namespace, monitor.Name, err)
+			return false, err
+		}
+		// If enable persistent,operator need to migrate pvc and pv binding relationship.
+		return !monitor.Spec.Persistent, nil
+	}
+
+	if !errors.IsNotFound(err) {
+		klog.Errorf("Fail to get deployment for tm [%s/%s], err: %v", monitor.Namespace, monitor.Name, err)
+		return false, err
+	}
+	if !monitor.Spec.Persistent {
+		return true, nil
+	}
+	firstStsPvcName := GetMonitorFirstPVCName(monitor.Name)
+	deploymentPvcName := GetMonitorObjectName(monitor)
+	deploymentPvc, err := m.deps.PVCLister.PersistentVolumeClaims(monitor.Namespace).Get(deploymentPvcName)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			klog.Errorf("Smooth migration for tm[%s/%s], get the PVC of the deployment error: %v", monitor.Namespace, monitor.Name, err)
+			return false, err
+		}
+
+		// If the PVC of the deployment does not exist and no old PV status, we don't need to migrate.
+		if monitor.Status.DeploymentStorageStatus == nil || len(monitor.Status.DeploymentStorageStatus.PvName) <= 0 {
+			return true, nil
+		}
+
+		deploymentPv, err := m.deps.PVLister.Get(monitor.Status.DeploymentStorageStatus.PvName)
+		if err != nil {
+			klog.Errorf("Smooth migration for tm[%s/%s], fail to patch PV %s, err: %v", monitor.Namespace, monitor.Name, monitor.Status.DeploymentStorageStatus.PvName, err)
+			return false, err
+		}
+
+		if deploymentPv.Spec.ClaimRef != nil && deploymentPv.Spec.ClaimRef.Name == firstStsPvcName {
+			// smooth migration successfully and clean status
+			monitor.Status.DeploymentStorageStatus = nil
+			return true, nil
+		}
+
+		err = m.patchPVClaimRef(deploymentPv, firstStsPvcName, monitor)
+		if err != nil {
+			klog.Errorf("Smooth migration for tm[%s/%s], fail to patch PV %s, err: %v", monitor.Namespace, monitor.Name, monitor.Status.DeploymentStorageStatus.PvName, err)
+			return false, err
+		}
+		// smooth migration successfully and clean status
+		monitor.Status.DeploymentStorageStatus = nil
+		return true, nil
+	}
+
+	if len(deploymentPvc.Spec.VolumeName) <= 0 {
```

Comment:
For this case, no need to migrate, we should print info log and `return true, nil`.